### PR TITLE
Coresight: add init coresight driver framework

### PR DESCRIPTION
**Pull request by leo11261**
## Summary
add init coresight driver framework

## Impact
debug features

## Testing
ETM->FUNNEL->ETF->ETR->psram
STM->FUNNEL->ETF->ETR->psram

coresight devices above are tested, and can produce trace data.

### CODE REVIEW

File: drivers/coresight/coresight_core.c [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: coresight_build_path
+ *
+ * Description:
+ *   Build path from srcdev to destdev.
+ *
```
Comment (acassis) [inline drivers/coresight/coresight_core.c:446]: Please include the input parameters

File: include/nuttx/coresight/coresight.h [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+ *   Disable the trace start from srcdev to destdev.
+ *
+ * Input Parameters:
+ *   srcdev  - Source device that generates trace data.
+ *
+ ****************************************************************************/
+
+void coresight_disable(FAR struct coresight_dev_s *srcdev);
+
+#endif  //__INCLUDE_NUTTX_CORESIGHT_CORESIGHT_H
```
Comment (acassis) [inline include/nuttx/coresight/coresight.h:279]: Please use /* ... */

File: drivers/coresight/coresight_common.c [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+                      addr + CORESIGHT_CLAIMCLR);
+    }
+
+out:
+  coresight_lock(addr);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: coresight_disclaim_device
```
Comment (acassis) [inline drivers/coresight/coresight_common.c:140]: Please include a complete function description comment

File: drivers/coresight/coresight_common.c [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  coresight_lock(addr);
+}
+
+/****************************************************************************
+ * Name: coresight_get_cpu_trace_id
+ *
+ * Description:
+ *   Used to get an unique trace id associated with cpu id of an ETM
+ *   coresight device.
+ *
```
Comment (acassis) [inline drivers/coresight/coresight_common.c:166]: Please include input parameter 'cpu'

File: drivers/coresight/coresight_common.c [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+  __set_bit(traceid, g_coresight_trace_id_bitmap);
+  spin_unlock_irqrestore(&g_coresight_trace_id_lock, flags);
+
+  return traceid;
+}
+
+/****************************************************************************
+ * Name: coresight_put_system_trace_id
+ *
```
Comment (acassis) [inline drivers/coresight/coresight_common.c:230]: Ditto: traceid

File: drivers/coresight/coresight_common.c [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+  irqstate_t flags;
+
+  flags = spin_lock_irqsave(&g_coresight_trace_id_lock);
+  __clear_bit(traceid, g_coresight_trace_id_bitmap);
+  spin_unlock_irqrestore(&g_coresight_trace_id_lock, flags);
+}
+
+/****************************************************************************
+ * Name: coresight_timeout
+ *
```
Comment (acassis) [inline drivers/coresight/coresight_common.c:247]: Ditto: please include input parameters

File: drivers/coresight/coresight_common.c [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+      up_udelay(1);
+    }
+
+  return -EAGAIN;
+}
+
+/****************************************************************************
+ * Name: coresight_insert_barrier_packet
+ *
```
Comment (acassis) [inline drivers/coresight/coresight_common.c:273]: Ditto: buf

**Review by acassis — CHANGES_REQUESTED** [review]
Please include some basic documentation talking about these debug features: "how, when and why" it should be used. I saw ARM has a good documentation about it: https://developer.arm.com/documentation/102520/latest/ so maybe you could include it as reference and explain which benefits this support on NuttX brings to user and how they can use it for everyday debugging, etc.

**Comment by acassis** [comment]
@leo11261 any news?

**Comment by leo11261** [comment]
> @leo11261 any news?

i will update them in few days.

**Comment by leo11261** [comment]
> Please include some basic documentation talking about these debug features: "how, when and why" it should be used. I saw ARM has a good documentation about it: https://developer.arm.com/documentation/102520/latest/ so maybe you could include it as reference and explain which benefits this support on NuttX brings to user and how they can use it for everyday debugging, etc.

i have update a document, refers to 

> Please include some basic documentation talking about these debug features: "how, when and why" it should be used. I saw ARM has a good documentation about it: https://developer.arm.com/documentation/102520/latest/ so maybe you could include it as reference and explain which benefits this support on NuttX brings to user and how they can use it for everyday debugging, etc.

refers to https://github.com/apache/nuttx/pull/11605/commits/f23733042b036fd58a75a9513197ad991d4b5e8b

**Comment by leo11261** [comment]
> Please include some basic documentation talking about these debug features: "how, when and why" it should be used. I saw ARM has a good documentation about it: https://developer.arm.com/documentation/102520/latest/ so maybe you could include it as reference and explain which benefits this support on NuttX brings to user and how they can use it for everyday debugging, etc.

I have already added one in latest commit，please refer to https://github.com/apache/nuttx/commit/6747f57fed7d59675a24da7d5f9c9411fc7fd91b

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**